nifi-nar-bundles/nifi-splunk-bundle/nifi-splunk-processors/src/main/java/org/apache/nifi/processors/splunk/ListenSplunkForwarder.java

```diff
@@ -0,0 +1,208 @@
+            .identifiesControllerService(SSLContextService.class)
+            .build();
+
+    // it is only the array reference that is volatile - not the contents.
+    private volatile byte[] messageDemarcatorBytes;
+
+    @Override
+    protected List<PropertyDescriptor> getAdditionalProperties() {
+        return Arrays.asList(
+                PROTOCOL,
+                MAX_CONNECTIONS,
+                MAX_BATCH_SIZE,
+                MESSAGE_DELIMITER,
+                SSL_CONTEXT_SERVICE
+        );
+    }
+
+    @Override
+    @OnScheduled
+    public void onScheduled(ProcessContext context) throws IOException {
+        super.onScheduled(context);
+        final String msgDemarcator = context.getProperty(MESSAGE_DELIMITER).getValue().replace("\\n", "\n").replace("\\r", "\r").replace("\\t", "\t");
+        messageDemarcatorBytes = msgDemarcator.getBytes(charset);
+    }
+
+    @Override
+    protected ChannelDispatcher createDispatcher(final ProcessContext context, final BlockingQueue<SplunkEvent> events)
+            throws IOException {
+
+        final String protocol = context.getProperty(PROTOCOL).getValue();
+        final int maxConnections = context.getProperty(MAX_CONNECTIONS).asInteger();
+        final int bufferSize = context.getProperty(RECV_BUFFER_SIZE).asDataSize(DataUnit.B).intValue();
+        final Charset charSet = Charset.forName(context.getProperty(CHARSET).getValue());
+
+        // initialize the buffer pool based on max number of connections and the buffer size
+        final LinkedBlockingQueue<ByteBuffer> bufferPool = new LinkedBlockingQueue<>(maxConnections);
+        for (int i = 0; i < maxConnections; i++) {
+            bufferPool.offer(ByteBuffer.allocate(bufferSize));
+        }
+
+        final EventFactory<SplunkEvent> eventFactory = new SplunkEventFactory();
+
+        if (UDP_VALUE.getValue().equals(protocol)) {
+            return new DatagramChannelDispatcher(eventFactory, bufferPool, events, getLogger());
+        } else {
+            // if an SSLContextService was provided then create an SSLContext to pass down to the dispatcher
+            SSLContext sslContext = null;
+            final SSLContextService sslContextService = context.getProperty(SSL_CONTEXT_SERVICE).asControllerService(SSLContextService.class);
+            if (sslContextService != null) {
+                sslContext = sslContextService.createSSLContext(SSLContextService.ClientAuth.REQUIRED);
+            }
+
+            final ChannelHandlerFactory<SplunkEvent<SocketChannel>, AsyncChannelDispatcher> handlerFactory = new SocketChannelHandlerFactory<>();
+            return new SocketChannelDispatcher(eventFactory, handlerFactory, bufferPool, events, getLogger(), maxConnections, sslContext, charSet);
+        }
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        final int maxBatchSize = context.getProperty(MAX_BATCH_SIZE).asInteger();
```

Comment:
Considering that this is using the similar batching approach used in PutKafka, curious if you tested it with run-duration > 0. See comments in https://issues.apache.org/jira/browse/NIFI-1534 as it may be suffering from the same issue.